### DISCUSSION TITLE
PointStruct is very slow

### DISCUSSION THREAD

**Discussion by dacox** [qdrant/qdrant-client#469] [General]
At first I thought the bottleneck in my indexing code was data access, but it turns out that it is actually the `python_client` `PointStruct`.

```
from qdrant_client.http.models import PointStruct

def f(embeddings):
    points = []
    for embedding in embeddings:
        points.append({
            'id': embedding.id,
            'vector': embedding.vector,
            'payload': {
                "asset_uuid": embedding.asset_uuid,
                "type": embedding.get_type(),
                "source": embedding.get_source(),
            },
        })
    return points

def g(embeddings):
    points = []
    for embedding in embeddings:
        points.append(PointStruct(
            id=embedding.id,
            vector=embedding.vector,
            payload={
                "asset_uuid": embedding.asset_uuid,
                "type": embedding.get_type(),
                "source": embedding.get_source(),
            },
        ))
    return points
```

Here are the timing comparisons:

```
%timeit _ = f(embeddings)
12.8 ms ± 243 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

%timeit _ = g(embeddings)
2.02 s ± 48.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

It's over 150x slower!

**Comment by joein** [comment]
Hi, could you please check which pydantic version are you using?

**Comment by dacox** [comment]
> Hi, could you please check which pydantic version are you using?

@joein 

Looking at our `poetry.lock`, it seems to be

```
[[package]]
name = "pydantic"
version = "1.10.13"
description = "Data validation and settings management using python type hints"
optional = false
python-versions = ">=3.7"
```

**Comment by joein** [comment]
pydantic v1 is indeed very slow, however I would not expect it to be this slow. Could you install pydantic v2 or do you have some blocking dependency?

**Comment by dacox** [comment]
@joein looking at the lockfile, its not immediately clear to me why were on v1. Will try with v2 and report back

**Comment by dacox** [comment]
@joein ok, i ran some tests with v2. I'm seeing `g` execute in more like `100ms` now. 

However, `f` is down to `2ms` 😆  So its still 50x slower, but I'm not really sure how to explain the speedup of the original method after upgrading `pydantic`

**Comment by joein** [comment]
serialization and deserialization in pydantic v2 is written in rust

I think it is the cost of data validation

If you don't want to validate your data in runtime, you can use `PointStruct.construct`, then you will have static type checking in your IDE and no runtime validation

**Comment by joein** [comment]
Do you mind if I convert this issue into a discussion?

**Comment by dacox** [comment]
> Do you mind if I convert this issue into a discussion?

go for it!
